client/client.go

```diff
@@ -249,6 +259,44 @@ func (c *client) updateTSODispatcher() {
 	})
 }
 
+func (c *client) leaderCheckLoop() {
+	defer c.wg.Done()
+
+	leaderCheckLoopCtx, leaderCheckLoopCancel := context.WithCancel(c.ctx)
+	defer leaderCheckLoopCancel()
+
+	ticker := time.NewTicker(leaderHealthCheckInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-ticker.C:
+			c.checkLeaderHealth(leaderCheckLoopCtx)
+		}
+	}
+}
+
+func (c *client) checkLeaderHealth(ctx context.Context) {
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+	if cc, ok := c.clientConns.Load(c.GetLeaderAddr()); ok {
+		healthCli := healthpb.NewHealthClient(cc.(*grpc.ClientConn))
+		resp, err := healthCli.Check(ctx, &healthpb.HealthCheckRequest{Service: ""})
+		rpcErr, ok := status.FromError(err)
+		failpoint.Inject("unreachableNetwork1", func() {
+			resp = nil
+			err = status.New(codes.Unavailable, "unavailable").Err()
+		})
+		if (ok && isNetworkError(rpcErr.Code())) || resp.GetStatus() != healthpb.HealthCheckResponse_SERVING {
```

Comment:
If the request timeout, such as the firewall dropping the packet, Will it be network failure?